src/runtime/virtcontainers/utils/utils_linux.go

```diff
@@ -152,3 +154,50 @@ func IsAPVFIOMediatedDevice(sysfsdev string) bool {
 	}
 	return false
 }
+
+func waitProcessUsingPidfd(pid int, timeoutSecs uint, logger *logrus.Entry) (bool, error) {
+	pidfd, err := unix.PidfdOpen(pid, 0)
+
+	if err != nil {
+		if err == unix.ESRCH {
+			return false, nil
+		}
+
+		return true, err
+	}
+
+	defer unix.Close(pidfd)
+
+	start := time.Now()
+	var n int
+
+	for {
+		remaining := int(timeoutSecs*1000) - int(time.Since(start).Milliseconds())
+		if remaining < 0 {
+			remaining = 0
+		}
+
+		n, err = unix.Poll([]unix.PollFd{{Fd: int32(pidfd), Events: unix.POLLIN}}, remaining)
+		if err != unix.EINTR {
+			break
+		}
+	}
+
+	if err != nil || n != 1 {
+		logger.Warnf("process %v still running after waiting %ds", pid, timeoutSecs)
+		return true, err
+	}
+
+	reap(pid)
```

Comment:
Let's go pidfd all the way and use a variant based on Waitid(P_PIDFD) instead of Wait4() ?

Reply:
Correct, modified to use pidfd for waiting.

Reply:
On further analysis i changed it back to wait4,  It looks like pidfd_open is introduced during kernel 5.3 and waitid with P_PIDFD is introduced in 5.4. For versions in-between waitid will return EINVAL, which is not what we want.

Reply:
In the end, I finally settled for waitid with P_PIDFD and if it returns EINVAL it will fall back to waitid with P_PID